core/kernel/dt.c

```diff
@@ -51,6 +51,93 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
 	return prop;
 }
 
+int dt_get_irq(void *fdt, int node)
+{
+	const uint32_t *int_prop = NULL;
+	int len_prop = 0;
+	int it_num = DT_INFO_INVALID_INTERRUPT;
+
+	/*
+	 * Interrupt property can be defined with at least 2x32 bits word
+	 *  - Type of interrupt
+	 *  - Interrupt Number
+	 */
+	int_prop = fdt_getprop(fdt, node, "interrupts", &len_prop);
+
+	if (!int_prop || (len_prop < 2))
+		return it_num;
+
+	it_num = fdt32_to_cpu(int_prop[1]);
+
+	return it_num;
+}
+
+int dt_find_node_by_compatible_and_status(void *fdt, int status, int node,
+						const char *compat)
+{
+	int fnode = -FDT_ERR_NOTFOUND;
+
+	fnode = fdt_node_offset_by_compatible(fdt, node, compat);
+
+	while (fnode != -FDT_ERR_NOTFOUND) {
+		if (_fdt_get_status(fdt, fnode) & status)
```

Comment:
@etienne-lms 
I'm sorry but I don't understand. 
The problem is not the `status` but the `_fdt_get_status` that returns `DT_STATUS_NSEC | DT_STATUS_SEC` if `status = okay` and `secure-status = okay`.
As properties are defined in the DTB, we can not control that. What the function is doing is to check if `_fdt_get_status` returned value as a `status` bit set or not.

Reply:
I think one could legitimately call this function with status=`DT_STATUS_DISABLED` to find a compatible node where both status and secure-status are disabled. Your implementation suggests that this is not a valid scenario.

Do you expect this function to find a node with exclusively one of `status` and `secure-status` being enabled, the other being disable, and never with both enable?

I still think a simpler, explicit and generic way is to expect target node get `_fdt_get_status()` to exactly match input argument `status`.

Or if there is no clean generic way, maybe move this function as a local function of your driver that needs it.

Reply:
You're right on the fact that the function is not handling the DT_STATUS_DISABLED. This is no so complex to add this case (I just didn't think about it).

Next for me the biggest problem is coming from the `_fdt_get_status` function itself in which I don't see why the `DT_STATUS_SEC` is returned because `status=okay` and `secure-status` is not present. This is a blocking point because their is no possibility to distinguish if the node is only secure.
In same cases, we can not considered that a node is secure because `status=okay`, some node should be exclusively Secure or Non-Secure.

See that we are discussing since a long time on this function and if you agree I'll move it back to my specific driver as it was before. Do you agree?

Reply:
> Next for me the biggest problem is coming from the _fdt_get_status function itself in which I don't see why the DT_STATUS_SEC is returned because status=okay and secure-status is not present. This is a blocking point because their is no possibility to distinguish if the node is only secure.

This behavior follows the DT bindings. No `status` is also equivalent to `status = "okay"`. A node enabled only for secure world should explicitly state:
```
    status = "disabled"; 
    secure-status = "okay";
```

Reply:
remove this function from generic code